src/category_theory/localization/construction.lean

```diff
@@ -0,0 +1,194 @@
+import category_theory.category.Quiv
+
+/-!
+
+# Construction of the localized category
+
+This file constructs the localized category, obtained by formally inverting
+a class of maps `W : arrow_class C` in a category `C`.
+
+We first construct a quiver `loc_quiver W` whose objects are the same as those
+of `C` and whose maps are the maps in `C` and placeholders for the formal
+inverses of the maps in `W`.
+
+The localizated category `localization W` is obtained by taking the quotient
+of the path category of `loc_quiver W` by the congruence generated by four
+types of relations.
+
+The obvious functor `Q W : C ⥤ localization W` satisfies the universal property
+of the localization. Indeed, if `G : C ⥤ D` sends morphisms in `W` to isomorphisms
+in `D` (i.e. we have `hG : W.is_inverted_by G`), then there exists a unique functor
+`G' : localization W ⥤ D` such that `Q W ≫ G' = G`. This `G'` is `lift G hG`.
+The expected property of `lift G hG` if expressed by the lemma `fac` and the
+uniqueness is expressed by `fac`.
+
+## References
+
+* [P. Gabriel, M. Zisman, *Calculus of fractions and homotopy theory*][gabriel-zisman-1967]
+
+-/
+
+noncomputable theory
+
+open category_theory.category
+
+namespace category_theory
+
+namespace arrow_class
+
+variables {C : Type*} [category C] (W : arrow_class C)
+variables {D : Type*} [category D]
+
+namespace localization
+
+/-- If `W : arrow_class C`, `loc_quiver W` is a quiver with the same objects as `C`, and
+whose morphisms are those in `C` and placeholders for formal inverses of the morphisms
+in `W`. -/
+@[nolint has_inhabited_instance]
+structure loc_quiver (W : arrow_class C) := (obj : C)
+
+instance : quiver (loc_quiver W) :=
+{ hom := λ A B, (A.obj ⟶ B.obj) ⊕ { f : B.obj ⟶ A.obj // arrow.mk f ∈ W} }
+
+/-- The object in the path category of `loc_quiver W` attached to an object in
+the category `C` such that `W : arrow_class C` -/
+def ι_paths (X : C) : paths (loc_quiver W) := ⟨X⟩
+
+/-- The morphism in the path category associated to a morphism in the original category. -/
+@[simp]
+def ψ₁' {X Y : paths (loc_quiver W)} (f : X.obj ⟶ Y.obj) : X ⟶ Y := paths.of.map (sum.inl f)
+
+/-- The morphism in the path category associated to a morphism in the original category. -/
+@[simp]
+def ψ₁ {X Y : C} (f : X ⟶ Y) : ι_paths W X ⟶ ι_paths W Y := paths.of.map (sum.inl f)
+
+/-- The morphism in the path category corresponding to a formal inverse. -/
+@[simp]
+def ψ₂' {X Y : paths (loc_quiver W)} (w : X.obj ⟶ Y.obj) (hw : arrow.mk w ∈ W) : Y ⟶ X :=
+paths.of.map (sum.inr ⟨w, hw⟩)
+
+/-- The morphism in the path category corresponding to a formal inverse. -/
+@[simp]
+def ψ₂ {X Y : C} (w : X ⟶ Y) (hw : arrow.mk w ∈ W) : ι_paths W Y ⟶ ι_paths W X := ψ₂' W w hw
+
+/-- The relations by which we take the quotient in order to get the localized category. -/
+inductive relations ⦃X Y : (paths (loc_quiver W))⦄ : (X ⟶ Y) → (X ⟶ Y) → Prop
+| id (φ : X ⟶ Y) :
+  relations (ψ₁' W (𝟙 X.obj) ≫ φ) φ
```

Comment:
Is there a reason for putting `X` and `Y` to the left of the colon? If they were on the right, I think you could simplify the constructors a little (e.g. the `φ` argument to `id` seems redundant)

Reply:
Thanks very much! This was only because I am not expert with inductive types... I have done the change in the last commit. This makes the code much cleaner.

Reply:
Looks great!